fix: Server side diff now works correctly with some fields removal

andrii-korotkov-verkada opened the pull request:
Helps with https://github.com/argoproj/argo-cd/issues/20792

Removed and modified sets may only contain the fields that changed, not including key fields like "name". This can cause merge to fail, since it expects those fields to be present if they are present in the predicted live. Fortunately, structured merge diff folks have added an option to extracting items to add key fields, and we are going to use that.

Failure of the new test before the fix
```
            	Error:      	Received unexpected error:
            	            	error removing non config mutations for resource Deployment/nginx-deployment: error reverting webhook removed fields in predicted live resource: .spec.template.spec.containers: element 0: associative list with keys has an element that omits key field "name" (and doesn't have default value)
            	Test:       	TestServerSideDiff/will_test_removing_some_field_with_undoing_changes_done_by_webhook
```

crenshaw-dev commented on pkg/diff/diff_test.go:
```diff
@@ -871,7 +871,7 @@ func TestStructuredMergeDiff(t *testing.T) {
 		require.NoError(t, err)
 		assert.NotNil(t, result)
-		assert.False(t, result.Modified)
+		assert.True(t, result.Modified)
```
Why does this result in a diff when it didn't before?

andrii-korotkov-verkada replied:
I've updated the deployment to actually have some changes, to better replicate the behavior from the bug report.

crenshaw-dev commented on pkg/diff/diff.go:
```diff
@@ -283,6 +285,28 @@ func removeWebhookMutation(predictedLive, live *unstructured.Unstructured, gvkPa
 	return &unstructured.Unstructured{Object: pl}, nil
 }
 
+// Appends the key fields like "name" to the set elements which have some entries keyed by those fields.
+// This is needed to merge properly, see https://github.com/argoproj/argo-cd/issues/20792.
+func appendKeyFields(set *fieldpath.Set) {
```
```suggestion
// appendKeyFields appends the key fields like "name" to the set elements which have some entries keyed by those fields.
// This is needed to merge properly, see https://github.com/argoproj/argo-cd/issues/20792.
func appendKeyFields(set *fieldpath.Set) {
```

andrii-korotkov-verkada replied:
Updating

crenshaw-dev commented on pkg/diff/diff.go:
```diff
@@ -283,6 +285,28 @@ func removeWebhookMutation(predictedLive, live *unstructured.Unstructured, gvkPa
+// Appends the key fields like "name" to the set elements which have some entries keyed by those fields.
+// This is needed to merge properly, see https://github.com/argoproj/argo-cd/issues/20792.
+func appendKeyFields(set *fieldpath.Set) {
+	// Map is to deduplicate
```
Can you expand the comment to explain why de-duplication is necessary? What goes wrong if we don't de-dupe?

andrii-korotkov-verkada replied:
I ended up replacing with a list, since a small optimization of reducing a number of calls to Has below might be overshadowed by the extra cost for a string conversion.

crenshaw-dev commented on pkg/diff/diff.go:
```diff
@@ -283,6 +285,28 @@ func removeWebhookMutation(predictedLive, live *unstructured.Unstructured, gvkPa
 
+// Appends the key fields like "name" to the set elements which have some entries keyed by those fields.
+// This is needed to merge properly, see https://github.com/argoproj/argo-cd/issues/20792.
+func appendKeyFields(set *fieldpath.Set) {
```
Is there a good way to unit test this function? I understand it's covered by higher-level function tests, but I think some tests specific to this function would really clarify how it works.

andrii-korotkov-verkada replied:
I'll look into this

andrii-korotkov-verkada replied:
Glad I did a test - it caught a subtle bug.

crenshaw-dev commented on pkg/diff/diff.go:
```diff
@@ -258,6 +258,7 @@ func removeWebhookMutation(predictedLive, live *unstructured.Unstructured, gvkPa
 	}
 
 	if comparison.Modified != nil && !comparison.Modified.Empty() {
```
Is this bug also potentially an issue for added fields? If not, why not? Maybe should add a comment to explain.

andrii-korotkov-verkada replied:
We don't call merge, so maybe not, but I'll check.

crenshaw-dev reviewed (changes requested):
Can you open an Argo CD PR to test this? Is there a way to test the higher-level bug in Argo CD so we have coverage of both the gitops-engine behavior and of the user-facing bug?

andrii-korotkov-verkada commented:
> Can you open an Argo CD PR to test this? Is there a way to test the higher-level bug in Argo CD so we have coverage of both the gitops-engine behavior and of the user-facing bug?

I'll do the testing.

leoluz commented on pkg/diff/diff.go:
```diff
@@ -283,6 +285,29 @@ func removeWebhookMutation(predictedLive, live *unstructured.Unstructured, gvkPa
+				for _, keyField := range *path[i].Key {
+					pathPartCopy := make([]fieldpath.PathElement, i+1)
+					copy(pathPartCopy, path[:i+1])
+					newPath := append(pathPartCopy, fieldpath.PathElement{FieldName: &keyField.Name})
```
You are assuming that the field `name` is always the key in lists which can be true or false depending on the resource. We can't assume schema structure from this library and must rely on the provided `GvkParser`. The problem is likely cause by the fact that the `GvkParser` isn't populated with the necessary default values and list keys for the specific schema. It is strange, however that the error is being reported from a `Deployment` resource as this should contain the necessary schema info IRRC. 
Please investigate more and let me know your findings.

leoluz replied:
@andrii-korotkov-verkada I understand. My point is, how you determine what is the key that should be considered when merging 2 hypothetical lists?

andrii-korotkov-verkada replied:
All of the keys should be considered as far as I understand. Let me take a further look to see if we can derive from schema.

andrii-korotkov-verkada replied:
I also don't know if key fields are supposed to be returned by the diff, e.g. since they aren't necessarily removed. It's just we use those fields in a special way to do a merge, which has a different requirements that diff computation can have.

andrii-korotkov-verkada replied:
For example, we don't want key fields in added items, since we remove them in this logic. So not sure default behavior of the libs should be changed and it's on us to adjust, which I'm trying to do with this PR.

leoluz reviewed (changes requested):
Please check my comment.

crenshaw-dev commented on pkg/diff/testdata/smd-deploy-live.yaml:
```diff
@@ -120,7 +125,10 @@ spec:
             - containerPort: 80
               protocol: TCP
-          resources: {}
+          resources:
+            requests:
+              memory: 512Mi
+              cpu: 500m
```
Rather than change the existing test, should we create a new test and leave the old one in place?

andrii-korotkov-verkada replied:
Alright, I'll create a 2nd deployment manifests files.

crenshaw-dev commented on pkg/diff/testdata/smd-deploy-live.yaml:
```diff
@@ -120,7 +125,10 @@ spec:
           ports:
             - containerPort: 80
               protocol: TCP
```
In the case of both `ports` and `containers`, bot the live and the config yaml contain only one list item. I think we should probably do a test involving multiple list items, ideally with differing orders, to demonstrate that the key matching works correctly.

andrii-korotkov-verkada replied:
Okay

andrii-korotkov-verkada commented:
@crenshaw-dev, I've addressed your comments.

Code Analysis complains that configs are not perfect, but that's okay since they aren't deployed anywhere.

andrii-korotkov-verkada commented:
Per discussion with @leoluz, I'd spend more time to understand why the Structured Merge Diff library's Merge doesn't derive key fields from paths, since the info is available. I may submit a ticket to their community with a request to update the library.

andrii-korotkov-verkada commented:
I've filed https://github.com/kubernetes-sigs/structured-merge-diff/issues/273

leoluz commented:
Structured-merge-diff PR with the fix is now merged:
https://github.com/kubernetes-sigs/structured-merge-diff/pull/275

leoluz commented on sonar-project.properties:
```diff
@@ -0,0 +1,2 @@
+# Exclude test manifests from analysis
+sonar.kubernetes.exclusions=pkg/diff/testdata/**
```
This doesn't seem to be working

leoluz reviewed (approved):
LGTM
